Userland/Libraries/LibWeb/HTML/HTMLMeterElement.cpp

```diff
@@ -22,4 +26,185 @@ void HTMLMeterElement::initialize(JS::Realm& realm)
+    if (!maybe_low.has_value())
+        return min();
+    return clamp(maybe_low.value(), min(), max());
+}
+
+WebIDL::ExceptionOr<void> HTMLMeterElement::set_low(double value)
+{
+    TRY(set_attribute(HTML::AttributeNames::low, MUST(String::number(value))));
+    update_meter_value_element();
+    document().invalidate_layout();
+    return {};
+}
+
+// https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-high
+double HTMLMeterElement::high() const
+{
+    auto maybe_high_string = get_attribute(HTML::AttributeNames::high);
+    if (!maybe_high_string.has_value())
+        return max();
+    auto maybe_high = parse_floating_point_number(maybe_high_string.value());
+    if (!maybe_high.has_value())
+        return max();
+    return clamp(maybe_high.value(), min(), max());
+}
+
+WebIDL::ExceptionOr<void> HTMLMeterElement::set_high(double value)
+{
+    TRY(set_attribute(HTML::AttributeNames::high, MUST(String::number(value))));
+    update_meter_value_element();
+    document().invalidate_layout();
+    return {};
+}
+
+// https://html.spec.whatwg.org/multipage/form-elements.html#concept-meter-optimum
+double HTMLMeterElement::optimum() const
+{
+    auto maybe_optimum_string = get_attribute(HTML::AttributeNames::optimum);
+    if (!maybe_optimum_string.has_value())
+        return (max() + min()) / 2;
+    auto maybe_optimum = parse_floating_point_number(maybe_optimum_string.value());
+    if (!maybe_optimum.has_value())
+        return (max() + min()) / 2;
+    return clamp(maybe_optimum.value(), min(), max());
+}
+
+WebIDL::ExceptionOr<void> HTMLMeterElement::set_optimum(double value)
+{
+    TRY(set_attribute(HTML::AttributeNames::optimum, MUST(String::number(value))));
+    update_meter_value_element();
+    document().invalidate_layout();
+    return {};
+}
+
+void HTMLMeterElement::inserted()
+{
+    create_shadow_tree_if_needed();
+}
+
+void HTMLMeterElement::create_shadow_tree_if_needed()
+{
+    if (shadow_root_internal())
+        return;
+
+    auto shadow_root = heap().allocate<DOM::ShadowRoot>(realm(), document(), *this, Bindings::ShadowRootMode::Closed);
+    set_shadow_root(shadow_root);
+
+    auto meter_bar_element = heap().allocate<MeterBarElement>(realm(), document());
+    MUST(shadow_root->append_child(*meter_bar_element));
+
+    m_meter_value_element = heap().allocate<MeterValueElement>(realm(), document());
+    MUST(meter_bar_element->append_child(*m_meter_value_element));
+    update_meter_value_element();
+}
+
+void HTMLMeterElement::update_meter_value_element()
+{
+    if (!m_meter_value_element)
+        return;
+
+    // UA requirements for regions of the gauge:
+    // If the optimum point is equal to the low boundary or the high boundary, or anywhere in between them, then the region between the low and high boundaries of the gauge must be treated as the optimum region, and the low and high parts, if any, must be treated as suboptimal.
+    if (optimum() >= low() && optimum() <= high()) {
```

Comment:
Each of these methods does quite a lot of logic, including parsing numbers. I think it would be worth calling them once each, saving their values, and then comparing those instead.

(Or as a step further, maybe it's worth caching these float values as `m_value`, `m_low`, etc, and setting those when the attribute changes? A lot of the getter methods call the other getter methods so things get parsed multiple times anyway.)

Reply:
I worry about desync issues so for now I query the attrs once as you suggested